pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"slices"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+// FQDNDataServer is the server for the standalone DNS proxy grpc server
+// It is responsible for handling the FQDN mapping requests from the SDP
+// and sending the DNS Policy updates to the SDP.
+type FQDNDataServer struct {
+	pb.UnimplementedFQDNDataServer
+
+	// port is the port on which the standalone DNS proxy grpc server will run
+	port int
+
+	// grpcServer is the grpc server for the standalone DNS proxy
+	grpcServer *grpc.Server
+
+	endpointManager endpointmanager.EndpointManager
+
+	// updateOnDNSMsg is a function to update the DNS message in the cilium agent on receiving the FQDN mapping
+	updateOnDNSMsg messagehandler.DNSRequestHandler
+
+	// snapshotMutex is a mutex to protect the current state of the DNS rules
+	snapshotMutex lock.Mutex
+
+	// identityToIpMutex is a mutex to protect the current state of the identity to Ip mapping
+	identityToIpMutex lock.Mutex
+
+	// currentIdentityToIp is a map of the identity to list of Ips
+	currentIdentityToIp map[identity.NumericIdentity][]net.IP
+
+	// log is the logger for the FQDNDataServer
+	log *slog.Logger
+}
+
+var (
+	kaep = keepalive.EnforcementPolicy{
+		PermitWithoutStream: true, // Allow pings even when there are no active streams
+	}
+	kasp = keepalive.ServerParameters{
+		Time:    5 * time.Second, // Ping the client if it is idle for 5 seconds to ensure the connection is still active
+		Timeout: 1 * time.Second, // Wait 1 second for the ping ack before assuming the connection is dead
+	}
+)
+
+type PolicyUpdater interface {
+	UpdatePolicyRulesLocked(map[identity.NumericIdentity]policy.SelectorPolicy, bool) error
```

Comment:
Document where this can be called and what the assumptions are. The method name ends in "Locked" which seems to imply some locking, but what locking and why?

Reply:
Removed the lock